src/misc/unicode.cpp

```diff
@@ -0,0 +1,1160 @@
+                    break;
+                ++i;
+            }
+        };
+
+        // Retrieve code point
+        uint32_t code_point = unknown_character;
+
+        // Support code point needing up to 3 bytes to encode; this includes
+        // Latin, Greek, Cyrillic, Hebrew, Arabic, VGA charset symbols, etc.
+        // More bytes are needed mainly for historic scripts, emoji, etc.
+
+        if (GCC_UNLIKELY(byte_1 >= threshold_bytes_6))
+            // 6-byte code point (>= 31 bits), no support
+            advance(5);
+        else if (GCC_UNLIKELY(byte_1 >= threshold_bytes_5))
+            // 5-byte code point (>= 26 bits), no support
+            advance(4);
+        else if (GCC_UNLIKELY(byte_1 >= threshold_bytes_4))
+            // 4-byte code point (>= 21 bits), no support
+            advance(3);
+        else if (GCC_UNLIKELY(byte_1 >= threshold_bytes_3)) {
+            // 3-byte code point - decode 1st byte
+            code_point = static_cast<uint8_t>(byte_1 - threshold_bytes_3);
+            // Decode 2nd byte
+            code_point = code_point << 6;
+            if (byte_2 >= threshold_non_ascii && byte_2 < threshold_encoded) {
+                ++i;
+                code_point = code_point + byte_2 - threshold_non_ascii;
+            }
+            // Decode 3rd byte
+            code_point = code_point << 6;
+            if (byte_2 >= threshold_non_ascii && byte_2 < threshold_encoded &&
+                byte_3 >= threshold_non_ascii && byte_3 < threshold_encoded) {
+                ++i;
+                code_point = code_point + byte_3 - threshold_non_ascii;
+            }
+        } else if (GCC_UNLIKELY(byte_1 >= threshold_encoded)) {
+            // 2-byte code point - decode 1st byte
+            code_point = static_cast<uint8_t>(byte_1 - threshold_encoded);
+            // Decode 2nd byte
+            code_point = code_point << 6;
+            if (byte_2 >= threshold_non_ascii && byte_2 < threshold_encoded) {
+                ++i;
+                code_point = code_point + byte_2 - threshold_non_ascii;
+            }
+        } else if (byte_1 < threshold_non_ascii)
+            // 1-byte code point, ASCII compatible
+            code_point = byte_1;
+
+        str_out.push_back(static_cast<uint16_t>(code_point));
+    }
+}
+
+static void warn_code_point(const uint16_t code_point)
+{
+    static std::set<uint16_t> already_warned;
+    if (already_warned.count(code_point))
+        return;
+    already_warned.insert(code_point);
+    LOG_WARNING("UTF8: No fallback mapping for code point 0x%04x", code_point);
+}
+
+static void warn_code_page(const uint16_t code_page)
+{
+    static std::set<uint16_t> already_warned;
+    if (already_warned.count(code_page))
+        return;
+    already_warned.insert(code_page);
+    LOG_WARNING("UTF8: Requested unknown code page %d", code_page);
+}
+
+static void wide_to_code_page(const std::vector<uint16_t> &str_in,
```

Comment:
The conversion never fails:
- if code page does not contain a character matching the grapheme we want to convert, it looks among the ALIAS'es (defined in MAIN.TXT); the list of valid aliases for the given code page is created when code page is first used
- if it fails, it drops the combining marks and tries again
- if it still fails, it tries to convert it to 7-bit ASCII using definitions from ASCII.TXT
- if it still fails, it is changed into a question mark, and a warning log is written in a console; only one warning is written for one Unicode code point within a lifecycle, to prevent spam (see 'warn_code_point' function)

It is perfectly fine that we are trying to display a string which is not fully supported by the current code page; the job of the engine is to provide the best output possible - by dropping accents, replacing Latin Schwa with Cyrillic Schwa (and vice-versa), and so on.
It is also possible, that a code page dedicated to the given language does not contain all the characters needed - see for example Georgian, which can be used with code pages 59829 and 60853; the second one contains more letters, but preserves almost no box/table drawing characters.

Reply:
Similar to the utf8_to_wide function, as mentioned there are cases the conversion may not be done properly (having to be replaced with question marks etc), and it is better to to indicate this to the calling methods which will then decide by themselves how to handle such cases. There will be likely be helper functions like `wide_to_utf8` and `code_page_to_wide` for making conversions in the other way eventually, and it is probably better to make them work in a consistent manner.

Reply:
@Wengier , the `wide_to_code_page` function's return type (and it's parent caller's return type) both return `std::string`, which was a change to address this PR comment: https://github.com/dosbox-staging/dosbox-staging/pull/1884#discussion_r975709131

Given this comment thread originated asking for the `void-return` type to be changed, I would say this thread is now overcome by events and be set to resolved.

Reply:
@kcgen If I understood correctly the discussion on Discord, @Wengier wants to have an indication that UTF8 conversion didn't go smoothly, for future features.

I have just changed the API to return bool, in a separate commit.